File: core/infra/src/test/java/com/beat/infra/persistence/user/mapper/UsersPersistenceMapperTest.java
package com.beat.infra.persistence.user.mapper;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

import org.junit.jupiter.api.Test;

import com.beat.domain.user.model.Role;
import com.beat.domain.user.model.Users;
import com.beat.infra.persistence.user.entity.UsersJpaEntity;

class UsersPersistenceMapperTest {

	private final UsersPersistenceMapper mapper = new UsersPersistenceMapper();

	@Test
	void roundTripPreservesPersistedUserRole() {
		Users users = Users.rehydrate(11L, Role.ADMIN);

		Users roundTrip = mapper.toDomain(mapper.toEntity(users));

		assertAll(
			() -> assertEquals(11L, roundTrip.getId()),
			() -> assertEquals(Role.ADMIN, roundTrip.getRole())
		);
	}

	@Test
	void toEntityKeepsGeneratedIdNullForNewUser() {
		Users users = Users.createWithRole(Role.MEMBER);

		UsersJpaEntity entity = mapper.toEntity(users);

		assertAll(
			() -> assertNull(entity.getId()),
			() -> assertEquals(Role.MEMBER, entity.getRole())
		);
	}
}

File: core/infra/src/test/kotlin/com/beat/infra/persistence/user/mapper/UsersPersistenceMapperTest.kt
package com.beat.infra.persistence.user.mapper

import com.beat.domain.user.model.Role
import com.beat.domain.user.model.Users
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe

class UsersPersistenceMapperTest : FunSpec({
    val mapper = UsersPersistenceMapper()

    test("roundTripPreservesPersistedUserRole") {
        val users = Users.rehydrate(11L, Role.ADMIN)

        val roundTrip = mapper.toDomain(mapper.toEntity(users))

        roundTrip.id shouldBe 11L
        roundTrip.role shouldBe Role.ADMIN
    }

    test("toEntityKeepsGeneratedIdNullForNewUser") {
        val users = Users.createWithRole(Role.MEMBER)

        val entity = mapper.toEntity(users)

        entity.id shouldBe null
        entity.role shouldBe Role.MEMBER
    }
})
